pypeman/msgstore.py

```diff
@@ -204,16 +207,21 @@ async def search(self, start=0, count=10, order_by='timestamp', start_dt=None, e
             end_dt = dateutil.parser.isoparse(end_dt)
 
         result = []
-        for value in sorted(self.messages.values(), key=lambda x: x[sort_key], reverse=reverse):
-            if start_dt and value["timestamp"] < start_dt:
-                continue
-            if end_dt and value["timestamp"] > end_dt:
-                continue
+        pos = 0
+        values = (
+            val for val in self.messages.values()
+            if (not start_dt or val["timestamp"] >= start_dt)
+            and (not end_dt or val["timestamp"] <= end_dt)
+        )
+        for value in sorted(values, key=lambda x: x[sort_key], reverse=reverse):
             resp = dict(value)
             resp['message'] = Message.from_dict(resp['message'])
-
-            result.append(resp)
-        return result[start: start + count]
+            if start <= pos < start + count:
+                result.append(resp)
+            elif pos >= start + count:
+                break
+            pos += 1
+        return result
```

Comment:
perhaps

from itertools import islice

```
for value in islice(sorted(values, key=lambda x: x[sort_key], reverse=reverse), start, start+count):
```

or if you think the lines too long:
```
sorted_vals = sorted(values, key=lambda x: x[sort_key], reverse=reverse)
for value in islice(sorted_vals, start, start + count):
```

Reply:
changed